pkg/statsd/receiver.go

```diff
@@ -96,99 +83,24 @@ func (mr *MetricReceiver) Receive(ctx context.Context, c net.PacketConn) {
 			continue
 		}
 		// TODO consider updating counter for every N-th iteration to reduce contention
-		atomic.AddUint64(&mr.packetsReceived, uint64(packetCount))
-		atomic.StoreInt64(&mr.lastPacket, time.Now().UnixNano())
-		atomic.AddUint64(&mr.batchesRead, 1)
+		atomic.AddUint64(&dr.packetsReceived, uint64(packetCount))
+		atomic.StoreInt64(&dr.lastPacket, time.Now().UnixNano())
+		atomic.AddUint64(&dr.batchesRead, 1)
 		for i := 0; i < packetCount; i++ {
 			addr := messages[i].Addr
 			nbytes := messages[i].N
 			buf := messages[i].Buffers[0]
-			if err := mr.handlePacket(ctx, addr, buf[:nbytes]); err != nil {
-				if err == context.Canceled || err == context.DeadlineExceeded {
-					return
-				}
-				log.Warnf("Failed to handle packet: %v", err)
+			msg := make([]byte, nbytes)
+			copy(msg, buf)
```

Comment:
This is a big drawback of this approach 😕

Reply:
To overcome this to some degree we can use [`sync.Pool`](https://golang.org/pkg/sync/#Pool) to pool free message buffers for reuse. I think we do it for something already in this project.

Reply:
I originally handled this by passing the entire buffer in and having a return channel to pass it back when I was done.  It's a fixed pool of memory and no copies.

Reply:
Also if we go that route, I think we should pass the batch, not the datagram.

Reply:
- Using `sync.Pool` is non-blocking vs reading/writing from/to return channel.
- Re. passing the whole batch: I also thought about it. It seems work will be redistributed less fairly that way. However we save on channel interactions so throughput should improve in case of many batches. So +1.

Reply:
Right, because we're worried about the GC overhead? I'll look into buffer reuse.

Reply:
I think it's more the copying overhead?  I'm vaguely curious about the GC overhead, I think it's up there, but I suspect the channel overhead is the big one - once this is in, I'm (or if you want to, go for it) going to try adding a small local buffer on the decoders, so multiple metrics are sent to the channel at a time.

Reply:
> Right, because we're worried about the GC overhead? I'll look into buffer reuse.

Both GC and copying are bad.